### PR TITLE
use shared prefs as backup storage for device id

### PR DESCRIPTION
**Pull request by djih**

### CODE REVIEW

File: src/com/amplitude/api/AmplitudeClient.java [hunk 1]
```diff
@@ -359,13 +359,6 @@ public AmplitudeClient setDiagnosticEventMaxCount(int eventMaxCount) {
         return this;
     }
 
-    // this method should only be called from the background log thread
-    private void initializeDeviceInfo() {
```
Comment (djih) [inline src/com/amplitude/api/AmplitudeClient.java]: unused

File: test/com/amplitude/api/AmplitudeClientTest.java [hunk 2]
```diff
@@ -251,56 +262,6 @@ public void testIdentifyMultipleOperations() throws JSONException {
         assertTrue(Utils.compareJSONObjects(userProperties, expected));
     }
 
-    @Test
-    public void testReloadDeviceIdFromDatabase() {
```
Comment (djih) [inline test/com/amplitude/api/AmplitudeClientTest.java]: Moved these tests into `InitializeTest`

File: src/com/amplitude/api/Utils.java [hunk 3]
```diff
@@ -103,4 +106,20 @@ static String normalizeInstanceName(String instance) {
         }
         return instance.toLowerCase();
     }
+
+    static SharedPreferences getAmplitudeSharedPreferences(Context context, String instanceName) {
+        assert !isEmptyString(instanceName);
+        String prefName = Constants.PACKAGE_NAME + "." + instanceName + "." + context.getPackageName();
+        return context.getSharedPreferences(prefName, Context.MODE_PRIVATE | Context.MODE_MULTI_PROCESS);
```
Comment (djih) [inline src/com/amplitude/api/Utils.java]: I think the `Context.MODE_MULTI_PROCESS` syncs the shared preferences across multiple processes, but probably not completely necessary

Comment (djih) [reply src/com/amplitude/api/Utils.java]: This is also a single instance shared across all clients so we don't need to cache a pointer to it anywhere (or write our own wrapper)

File: src/com/amplitude/api/AmplitudeClient.java [hunk 4]
```diff
@@ -2087,28 +2080,45 @@ private String initializeDeviceId() {
...
+                Utils.writeStringToSharedPreferences(context, instanceName, DEVICE_ID_KEY, deviceId);
+            }
+
             return deviceId;
         }
 
+        // backup #1: check if device id is stored in shared preferences
+        if (!(Utils.isEmptyString(sharedPrefDeviceId) || invalidIds.contains(sharedPrefDeviceId))) {
+            dbHelper.insertOrReplaceKeyValue(DEVICE_ID_KEY, sharedPrefDeviceId);
+            return sharedPrefDeviceId;
```
Comment (paladin8) [inline src/com/amplitude/api/AmplitudeClient.java]: I think these two cases are correct, but I kind of like the idea of all paths of this method ending up calling `saveDeviceId` just so it's easy to verify that we never put ourselves in an inconsistent state.

File: src/com/amplitude/api/Utils.java [hunk 5]
```diff
@@ -103,4 +106,20 @@ static String normalizeInstanceName(String instance) {
         }
         return instance.toLowerCase();
     }
+
+    static SharedPreferences getAmplitudeSharedPreferences(Context context, String instanceName) {
+        assert !isEmptyString(instanceName);
+        String prefName = Constants.PACKAGE_NAME + "." + instanceName + "." + context.getPackageName();
```
Comment (paladin8) [inline src/com/amplitude/api/Utils.java]: I assume this is similar to how we scope the DB data?

File: src/com/amplitude/api/AmplitudeClient.java (hunk 4 above)
Comment (djih) [reply src/com/amplitude/api/AmplitudeClient.java]: Good suggestion

File: src/com/amplitude/api/Utils.java (hunk 5 above)
Comment (djih) [reply src/com/amplitude/api/Utils.java]: Ya this is how we name our db file: https://github.com/amplitude/Amplitude-Android/blob/master/src/com/amplitude/api/DatabaseHelper.java#L74

This is how we named our legacy shared prefs file: https://github.com/amplitude/Amplitude-Android/blob/master/src/com/amplitude/api/AmplitudeClient.java#L2201-L2219

**Pull request merged**